File: SextoSemestre/algs-avancados/03/exactsum.cpp
#include <iostream>
#include <algorithm>
#include <vector>

using namespace std;

vector<long long> livros(10000);

int main()
{
    long long N, M;

    while (cin >> N)
    {
        for (long long i = 0; i < N; ++i)
            cin >> livros[i];

        sort(livros.begin(), livros.begin() + N);

        cin >> M;

        long long i = 0, j = N - 1;
        long long bI, bJ;
        for (; i < j;)
        {
            if (livros[i] + livros[j] < M)
                ++i;
            else if (livros[i] + livros[j] == M)
            {
                bI = i;
                bJ = j;
                ++i;
                --j;
            }
            else
                --j;
        }

        cout << "Peter deve comprar os livros com preços " << livros[bI] << " e " << livros[bJ] << ".\n\n";
    }
}

File: SextoSemestre/algs-avancados/03/exactsum.java
import java.util.*;

public class Main {
    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);

        while (scanner.hasNext()) {
            int N = scanner.nextInt();
            ArrayList<Long> livros = new ArrayList<>();

            for (int i = 0; i < N; ++i) {
                livros.add(scanner.nextLong());
            }

            Collections.sort(livros);

            long M = scanner.nextLong();
            int i = 0, j = N - 1;
            int bI = 0, bJ = 0;

            while (i < j) {
                long soma = livros.get(i) + livros.get(j);
                if (soma < M) {
                    ++i;
                } else if (soma == M) {
                    bI = i;
                    bJ = j;
                    ++i;
                    --j;
                } else {
                    --j;
                }
            }

            System.out.println("Peter deve comprar os livros com preços " + livros.get(bI) + " e " + livros.get(bJ) + ".\n");
        }
    }
}
